### PR TITLE
Micro-XRCE use bin interface over xml to save flash space

### PR DESCRIPTION
**Pull request by PetervdPerk-NXP**
Reduces the amount XML used in the generated Micro XRCE DDS code

On PX4 FMUv5 RTPS it saves around 5KB of flash.

Needs some proper testing since I can't due that my ROS setup is currently migrating to Humble.

FYI @bperseghetti 

Before
```
Memory region         Used Size  Region Size  %age Used
      FLASH_ITCM:          0 GB      2016 KB      0.00%
      FLASH_AXIM:     1905757 B      2016 KB     92.32%
        ITCM_RAM:          0 GB        16 KB      0.00%
        DTCM_RAM:          0 GB       128 KB      0.00%
           SRAM1:       44712 B       368 KB     11.87%
           SRAM2:          0 GB        16 KB      0.00%
```
After
```
Memory region         Used Size  Region Size  %age Used
      FLASH_ITCM:          0 GB      2016 KB      0.00%
      FLASH_AXIM:     1900661 B      2016 KB     92.07%
        ITCM_RAM:          0 GB        16 KB      0.00%
        DTCM_RAM:          0 GB       128 KB      0.00%
           SRAM1:       44712 B       368 KB     11.87%
           SRAM2:          0 GB        16 KB      0.00%
```

### CODE REVIEW

**Review by TSC21 — APPROVED** [review]

**Review by bkueng — COMMENTED** [review]
Looks good, thanks @PetervdPerk-NXP. As this is conflicting with https://github.com/PX4/PX4-Autopilot/pull/19666, can you coordinate with @dsix-ls2n?

**Comment by PetervdPerk-NXP** [comment]
Backported @dsix-ls2n changes to this PR.
@bkueng if someone could test and verify this that would be great.

**Comment by PetervdPerk-NXP** [comment]
Furthermore regarding uxr_buffer_create_participant_xml
There's this _localhost_only option but I doubt it does anything did anyone verify this?
Looking at the DDS-XRCE participant api I think it processes the domain_id and participant name.

I don't see any mention of localhost_only, interfaceWhiteList, transport_id in the code.

**Comment by dsix-ls2n** [comment]
Hi, thanks for backporting #19666. However, I also added the namespace in the name of the participant initially. If I'm not wrong, this is not ported here. I think it is important for multiple instances on the same network (or computer in SITL).
I don't have the means right now to do proper testing either.

**Comment by PetervdPerk-NXP** [comment]
I don't think it's an hard requirement have different participant names.
What we also could is to use the PX4 GUID in the participant name.

**Review by bkueng — COMMENTED** [review]
> There's this _localhost_only option but I doubt it does anything did anyone verify this?

Yes, I've been using this.

I tested this quickly and I can receive topics in ROS, but in the other direction I don't receive data on the PX4 side.

File: msg/templates/uorb_microcdr/dds_topics.h.em [hunk 1]
```diff
@@ -37,6 +37,18 @@ receive_base_types = [s.short_name for idx, s in enumerate(spec) if scope[idx] =
 #include <uORB/ucdr/@(receive_base_types[idx]).h>
 @[end for]@
 
+#define TOPIC_NAME_SIZE 128
+
+static bool generate_topic_name(char* topic, const char* client_namespace, const char* direction, const char* name) {
+	if(client_namespace == nullptr) {
+		return snprintf(topic, TOPIC_NAME_SIZE, "rt/fmu/%s/%s",
+						direction, name) > 0;
```
[the remark this answers is not in the document]
Comment (PetervdPerk-NXP) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: I've pushed a commit with a check, any objections into merging this?

[the remark this answers is not in the document]
Comment (bkueng) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: > I tested this quickly and I can receive topics in ROS, but in the other direction I don't receive data on the PX4 side.

Does this work now?

[the remark this answers is not in the document]
Comment (PetervdPerk-NXP) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: Ah I see, I will try to setup the Micro-XRCE again on the companion computer.
By the way current PX4 user guide only mentions microRTPS.
Is there already a preliminary guide on how setup DDS-XRCE with PX4 on the Linux side?

[the remark this answers is not in the document]
Comment (bkueng) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: Not yet, it should be reasonably simple though. You build the agent (https://github.com/eProsima/Micro-XRCE-DDS-Agent) and then run it with `./MicroXRCEAgent udp4 localhost -p 15555` (if you use networking)

[the remark this answers is not in the document]
Comment (dsix-ls2n) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: Hi everyone. I wanted to know if there are some updates about this issue. What is missing to merge this?

[the remark this answers is not in the document]
Comment (dagar) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: @dsix-ls2n I have this working as part of a bigger cleanup branch that will be a pull request soon. https://github.com/PX4/PX4-Autopilot/tree/pr-ros2_uorb_rename_and_micrortps_purge

I was able to convert a few more things from xml -> bin and fix the incoming data (QoS issue with the DataReaders).

[the remark this answers is not in the document]
Comment (dsix-ls2n) [reply msg/templates/uorb_microcdr/dds_topics.h.em]: Perfect. Thank you this is great. So we might close this and #19666 ?

**Comment by dagar** [comment]
Rolled into https://github.com/PX4/PX4-Autopilot/pull/20227.

**Pull request closed without merging**